pkg/apply/filter/dependency-filter.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package filter
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cli-utils/pkg/apply/taskrunner"
+	"sigs.k8s.io/cli-utils/pkg/object"
+)
+
+//go:generate stringer -type=Action -linecomment
+type Action int
+
+const (
+	ApplyAction  Action = iota // Apply
+	DeleteAction               // Delete
+)
+
+// DependencyFilter implements ValidationFilter interface to determine if an
+// object can be applied or deleted based on the status of it's dependencies.
+type DependencyFilter struct {
+	TaskContext *taskrunner.TaskContext
+	Action      Action
+}
+
+const DependencyFilterName = "DependencyFilter"
+
+// Name returns the name of the filter for logs and events.
+func (dnrf DependencyFilter) Name() string {
+	return DependencyFilterName
+}
+
+// Filter returns true if the specified object should be skipped because at
+// least one of its dependencies is Not Found or Not Reconciled.
+func (dnrf DependencyFilter) Filter(obj *unstructured.Unstructured) (bool, string, error) {
+	id := object.UnstructuredToObjMetadata(obj)
+
+	switch dnrf.Action {
+	case ApplyAction:
+		edgesFrom := dnrf.TaskContext.Graph.EdgesFrom(id)
+		for _, depID := range edgesFrom {
+			if dnrf.TaskContext.IsInvalidObject(depID) {
```

Comment:
I have no idea if we really need to check all these statuses... Maybe we only need to check `IsSuccessfulReconcile`? But the error message would then be misleading saying it didn't reconcile when it was never applied...